kythe/docs/asciidoc.bzl

```diff
@@ -1,26 +1,75 @@
-def asciidoc(name, src, attrs={}, confs=[], data=[], tools=[], tags=None):
-  args = ['--backend', 'html', '--no-header-footer']
-  for key, val in attrs.items():
-    if val:
-      args += ["--attribute=%s=%s" % (key, val)]
-    else:
-      args += ["--attribute=%s!" % key]
-  args += ["--conf-file=$(location %s)" % c for c in confs]
-  out = name + '.html'
-  native.genrule(
-    name = name,
-    srcs = [src] + data + confs,
-    outs = [out],
-    local = True,
-    tags = tags,
-    tools = tools,
-    output_to_bindir = 1,
-    cmd = '\n'.join([
-        'export OUTDIR="$$PWD/$(@D)"',
-        'export LOGFILE="$$(mktemp -t \"XXXXXXasciidoc\")"',
-        'export PATH',
-        'trap "rm -f \"$${LOGFILE}\"" EXIT ERR INT',
-        "asciidoc %s -o $(@) $(location %s) 2> \"$${LOGFILE}\"" % (' '.join(args), src),
-        'cat $${LOGFILE}',
-        '! grep -q -e "filter non-zero exit code" -e "no output from filter" "$${LOGFILE}"']))
+def _asciidoc_impl(ctx):
+    asciidoc = ctx.toolchains["//tools/build_rules/external_tools:external_tools_toolchain_type"].asciidoc
+    args = ["--backend", "html", "--no-header-footer"]
```

Comment:
Use `ctx.actions.args()` rather than a list? https://docs.bazel.build/versions/master/skylark/lib/Args.html

Reply:
I think Args can't easily be turned back into a string list and passing it in as the arguments attribute means I'd have to rewrite the command to use $1, $2, etc. syntax to access the arguments. I didn't like how that looked so I kept it like this.

I could have just been holding it wrong though.